(msyang)-(~/test/gentoo-alt/tags/overlay/my-overlay/sci-libs/scipy/files)
(jobs:2) $ cat scipy-0.7.1-spatial-dickeyfuller.patch 
--- ./scipy/spatial/distance.py.orig 2009-08-22 04:48:08.000000000 -0400
+++ ./scipy/spatial/distance.py 2009-08-22 04:48:59.000000000 -0400
@@ -56,6 +56,8 @@
 +------------------+-------------------------------------------------+
 | dice             | the Dice dissimilarity (boolean).               |
 +------------------+-------------------------------------------------+
+| dickeyfuller     | the dickey-fuller statistic                     |
++------------------+-------------------------------------------------+
 | euclidean        | the Euclidean distance.                         |
 +------------------+-------------------------------------------------+
 | hamming          | the Hamming distance (boolean).                 |
@@ -841,6 +843,10 @@
     The following are common calling conventions.
+    0. ``Y = pdist(X, 'dickeyfuller')``
+
+       Computes the dickey-fuller 1-lag statistic.
+
     1. ``Y = pdist(X, 'euclidean')``
        Computes the distance between m points using Euclidean distance
@@ -1026,7 +1032,7 @@
        metric : string or function
            The distance metric to use. The distance function can
            be 'braycurtis', 'canberra', 'chebyshev', 'cityblock',
-           'correlation', 'cosine', 'dice', 'euclidean', 'hamming',
+           'correlation', 'cosine', 'dice', 'dickeyfuller', 'euclidean', 'hamming',
            'jaccard', 'kulsinski', 'mahalanobis', 'matching',
            'minkowski', 'rogerstanimoto', 'russellrao', 'seuclidean',
            'sokalmichener', 'sokalsneath', 'sqeuclidean', 'yule'.
@@ -1112,6 +1118,8 @@
         #    TypeError('A double array must be passed.')
         if mstr in set(['euclidean', 'euclid', 'eu', 'e']):
             _distance_wrap.pdist_euclidean_wrap(_convert_to_double(X), dm)
+        elif mstr in set(['dickeyfuller', 'dickey', 'df', 'd']):
+            _distance_wrap.pdist_df_wrap(_convert_to_double(X), dm)
         elif mstr in set(['sqeuclidean', 'sqe', 'sqeuclid']):
             _distance_wrap.pdist_euclidean_wrap(_convert_to_double(X), dm)
             dm = dm ** 2.0
--- ./scipy/spatial/src/distance.c.orig 2009-08-22 04:48:15.000000000 -0400
+++ ./scipy/spatial/src/distance.c 2009-08-22 04:49:26.000000000 -0400
@@ -39,6 +39,37 @@
 #include "common.h"
 #include "distance.h"
+static inline double df_distance(const double *u, const double *v, int n) {
+  int i = 0;
+  double mean_spread = 0.0, mean_nolag = 0.0, mean_lag = 0.0;
+  double result = 0.0;
+  double covar = 0.0;
+  // extra precision
+  //long double covar = 0.0;
+  for (i = 0; i < n; i++) {
+    const double spread = u[i] - v[i];
+    const double delta = spread - mean_spread;
+    mean_spread += delta/(i + 1);
+    result += delta * (spread - mean_spread);
+    if (i != 0) {
+        const double delta_lag = spread - mean_lag;
+        mean_lag += delta_lag/i;
+    }
+    if (i == n - 2){
+        mean_nolag = mean_spread;
+    }
+  }
+  for (i = 0; i < n - 1; i++) {
+      // extra precision
+      //const long double delta1 = u[i] - v[i] - mean_nolag;
+      //const long double delta2 = u[i+1] - v[i+1] - mean_lag;
+      const double delta1 = u[i] - v[i] - mean_nolag;
+      const double delta2 = u[i+1] - v[i+1] - mean_lag;
+      covar += (delta1 * delta2 - covar) / (i + 1);
+  }
+  return covar*(n - 1)/result;
+}
+
 static inline double euclidean_distance(const double *u, const double *v, int n) {
   int i = 0;
   double s = 0.0, d;
@@ -322,6 +353,19 @@
   }
 }
+void pdist_df(const double *X, double *dm, int m, int n) {
+  int i, j;
+  const double *u, *v;
+  double *it = dm;
+  for (i = 0; i < m; i++) {
+    for (j = i + 1; j < m; j++, it++) {
+      u = X + (n * i);
+      v = X + (n * j);
+      *it = df_distance(u, v, n);
+    }
+  }
+}
+
 void pdist_euclidean(const double *X, double *dm, int m, int n) {
   int i, j;
   const double *u, *v;
--- ./scipy/spatial/src/distance_wrap.c.orig 2009-08-22 04:48:25.000000000 -0400
+++ ./scipy/spatial/src/distance_wrap.c 2009-08-22 04:49:40.000000000 -0400
@@ -564,6 +564,27 @@
 /***************************** pdist ***/
+extern PyObject *pdist_df_wrap(PyObject *self, PyObject *args) {
+  PyArrayObject *X_, *dm_;
+  int m, n;
+  double *dm;
+  const double *X;
+  if (!PyArg_ParseTuple(args, "O!O!",
+   &PyArray_Type, &X_,
+   &PyArray_Type, &dm_)) {
+    return 0;
+  }
+  else {
+    X = (const double*)X_->data;
+    dm = (double*)dm_->data;
+    m = X_->dimensions[0];
+    n = X_->dimensions[1];
+
+    pdist_df(X, dm, m, n);
+  }
+  return Py_BuildValue("d", 0.0);
+}
